File: Flustered Pterodactyl/cocos2dx/gaf_runtimes/cocos2d-x/gaf/GAFTextureAtlas.cpp
#include "GAFTextureAtlas.h"
#include "GAFTextureAtlasElement.h"
#include "GAFData.h"
#include "platform/CCImage.h"
#include "textures/CCTexture2D.h"
#include "cocoa/CCArray.h"
#include "cocoa/CCDictionary.h"
#include "platform/CCFileUtils.h"

#if CC_ENABLE_CACHE_TEXTURE_DATA
#include "textures/CCTextureCache.h"
#endif

static const char * kElementsKey     = "elements";
static const char * kImagePathKey    = "source";

GAFTextureAtlas::GAFTextureAtlas()
:
_loaded(false),
_images(NULL),
_textures(NULL),
_elements(NULL)
{
}

GAFTextureAtlas::~GAFTextureAtlas()
{
	CC_SAFE_RELEASE(_images);
	CC_SAFE_RELEASE(_textures);
	CC_SAFE_RELEASE(_elements);
}

GAFTextureAtlas * GAFTextureAtlas::create(const char * jsonPath, CCDictionary * aConfigDictionary, CCArray * anImageDatas)
{
	GAFTextureAtlas *pRet = new GAFTextureAtlas();
    if (pRet && pRet->init(jsonPath, aConfigDictionary, anImageDatas))
    {
		pRet->autorelease();
        return pRet;
    }
    CC_SAFE_DELETE(pRet);
    return NULL;
}


bool GAFTextureAtlas::init(const char * jsonPath, CCDictionary * aConfigDictionary, CCArray * anImageDatas)
{
	if (!aConfigDictionary || !anImageDatas)
	{
		CCLOG("parameters should not equal to nil");
		return false;
	}
	
	CC_SAFE_RELEASE(_images);
	_images = new CCArray();
	
	for (int i = 0; i < anImageDatas->count(); ++i)
	{
		CCDictionary * data = (CCDictionary*)anImageDatas->objectAtIndex(i);
		CCString * source = (CCString *)data->objectForKey(kImagePathKey);
		CCImage * image = new CCImage();
		const char * path = CCFileUtils::sharedFileUtils()->fullPathFromRelativeFile(source->getCString(), jsonPath);
		image->initWithImageFile(path);
		_images->addObject(image);
		image->release();
	}
	loadElementsFromAnimationConfigDictionary(aConfigDictionary);
	return true;
}

bool GAFTextureAtlas::loadElementsFromAnimationConfigDictionary(CCDictionary * aConfigDictionary)
{
    CCArray   * nElements     = (CCArray*)   aConfigDictionary->objectForKey(kElementsKey);
	
	if (nElements)
	{
		int s = nElements->count();
		
		CC_SAFE_RELEASE(_elements);
		_elements = CCDictionary::create();
		CC_SAFE_RETAIN(_elements);
		
		for (int i = 0; i < s; ++i)
		{
			CCObject * nElement = nElements->objectAtIndex(i);
			CCDictionary * dict = dynamic_cast<CCDictionary*>(nElement);
			if (dict)
			{
				GAFTextureAtlasElement * element = GAFTextureAtlasElement::create(dict);
				if (element)
				{
					_elements->setObject(element, element->name);
				}
			}
			else
			{
				CCLOGERROR("Error when parsing texture atlas element JSON. Atlas element not of CCDictionary type");
			}
		}
	}
	else
	{
		CCLOGERROR("Error when parsing texture atlas element JSON.");
		return false;
	}
	_loaded = true;
	return true;
}


CCDictionary * GAFTextureAtlas::elements()
{
	return _elements;
}

CCImage     * GAFTextureAtlas::image()
{
	if (_images && _images->count() > 0)
	{
		return (CCImage*) _images->objectAtIndex(0);
	}
	return NULL;
}

CCArray     * GAFTextureAtlas::images()
{
	return _images;
}

CCTexture2D * GAFTextureAtlas::texture()
{
	if (_textures && _textures->count() > 0)
	{
		return (CCTexture2D*) _textures->objectAtIndex(0);
	}
	return NULL;
}

CCArray * GAFTextureAtlas::textures()
{
	if (!_textures)
	{
		_textures = CCArray::createWithCapacity(_images->count());
		for (int i = 0; i < _images->count(); ++i)
		{
			CCTexture2D * texture = new CCTexture2D();
			CCImage * image = (CCImage*)_images->objectAtIndex(i);
			texture->initWithImage(image);
			_textures->addObject(texture);
#if CC_ENABLE_CACHE_TEXTURE_DATA
			VolatileTexture::addCCImage(texture, image);
#endif
			texture->release();
		}
		_textures->retain();
	}
	return _textures;
}

File: BIBs/cocos2dx/gaf_runtimes/cocos2d-iphone/gaf/GAFTextureAtlas.m
#import "GAFTextureAtlas.h"
#import "GAFTextureAtlasElement.h"
#import "CCTexture2D.h"
#import "CCFileUtils.h"

static NSString * const kElementsKey     = @"elements";
static NSString * const kImagePathKey    = @"source";
static NSString * const kTextureAtlasKey = @"textureAtlas";

@interface GAFTextureAtlas ()

@property (nonatomic, retain) NSMutableArray *images;
@property (nonatomic, retain) NSMutableArray *textures;

- (void)loadElementsFromAnimationConfigDictionary:(NSDictionary *)aConfigDictionary;

@end

@implementation GAFTextureAtlas

#pragma mark -
#pragma mark Properties

- (CCTexture2D *)texture
{
    if (self.textures != nil && [self.textures count] > 0)
    {
        return self.textures[0];
    }
    return nil;
}

- (NSArray *)textures
{
    if (_textures == nil)
    {
        _textures = [[NSMutableArray alloc] initWithCapacity:self.images.count];
        
        for (UIImage *image in self.images)
        {
            CCTexture2D *texture = [[CCTexture2D alloc] initWithCGImage:image.CGImage resolutionType:kCCResolutioniPad];
            [_textures addObject:texture];
            [texture release];
        }
    }
    return _textures;
}

#pragma mark -
#pragma mark Initialization & Release

- (id)init
{
    self = [super init];
    if (nil != self)
    {
        self.images = [[[NSMutableArray alloc] initWithCapacity:3] autorelease];
        
        self.loaded = NO;
    }
    return self;
}


- (id)initWithJsonPath:(NSString*)jsonPath configDictionary:(NSDictionary *)aConfigDictionary imagesData:(NSArray *)anImageDatas
{
    NSAssert(anImageDatas != nil && aConfigDictionary != nil, @"parameters should not equal to nil");
    
    self = [self init];
	
	NSString * atlasDirectory = [jsonPath stringByDeletingLastPathComponent];
    if (nil != self)
    {
        for (NSDictionary *data in anImageDatas)
        {
			NSString * source = [data objectForKey:kImagePathKey];
            UIImage * image = [UIImage imageWithContentsOfFile:[atlasDirectory stringByAppendingPathComponent:source]]; // autoreleased object
            [_images addObject:image];
        }
        
        [self loadElementsFromAnimationConfigDictionary:aConfigDictionary];
    }
    return self;
}

- (void)dealloc
{
    [_images release];
    [_textures release];
    [_elements release];
    [super dealloc];
}

#pragma mark -
#pragma mark Public methods

#pragma mark -
#pragma mark Private methods

- (void)loadElementsFromAnimationConfigDictionary:(NSDictionary *)aConfigDictionary
{
    NSArray *nElements = [aConfigDictionary objectForKey:kElementsKey];
    
    if (nElements != nil)
    {
        @autoreleasepool
        {
            NSMutableDictionary *parsedElements = [[NSMutableDictionary alloc] initWithCapacity:[nElements count]];
            for (id nElement in nElements)
            {
                if ([nElement isKindOfClass:[NSDictionary class]])
                {
                    GAFTextureAtlasElement *element =
                        [[GAFTextureAtlasElement alloc] initWithDictionary:(NSDictionary *)nElement];
                    if (element != nil)
                    {
                        [parsedElements setObject:element forKey:element.name];
                        [element release];
                    }
                }
                else
                {
                   CCLOGWARN(@"Error when parsing texture atlas element JSON. Atlas element not of NSDictionary type");
                }
            }
            self.elements = parsedElements;
            [parsedElements release];
        }
    }
    else
    {
        CCLOGWARN(@"Error when parsing texture atlas element JSON.");
        return;
    }
    
    self.loaded = YES;
}

@end